WordPress/gutenberg
# [RNMobile] Add optional chaining to `reusableBlock.title` and `reusableBlock.content`

Potential fix for https://github.com/wordpress-mobile/gutenberg-mobile/issues/5496 and https://github.com/wordpress-mobile/WordPress-iOS/issues/21083

* Gutenberg Mobile PR: https://github.com/wordpress-mobile/gutenberg-mobile/pull/6228

## What?

This PR adds safety checks for the `title` and `content` properties within the `buildReusableBlockInserterItem` and `getUserPatterns` functions.

## Why?

We've received multiple reports of Gutenberg Mobile crashing with the following error:

```
TypeError: Cannot read property 'raw' of undefined
```

Even after digging deeply into user reports and available logs, attempts to reproduce the crash have been unsuccessful. The main clues we have to go on are as follows:

* The most recent logs vary, but usually reference `block-editor/src/store/selectors.js` or `block-editor/src/store/actions.js`.
* The only reference of `raw` in those files is within `block-editor/src/store/selectors.js`.
* While reviewing some logs, I noted that the crash occurs frequently while users are on pages with _unsynced_ patterns, so it feels logical that the reference to `raw` in `getUserPatterns` could be at fault.
* A crash with a similar log went away after adding safety checks to one of those references in https://github.com/WordPress/gutenberg/pull/53721. Though, it's worth noting we're not clear if the crash was resolved or simply "replaced" with this newer crash.

With this PR, we're making another educated guess at the root cause of the crash in an attempt to fix it.

## How?

Similar to https://github.com/WordPress/gutenberg/pull/53721, I'm proposing optional chaining be added in the three remaining places where `raw` appears without optional chaining in the `block-editor/src/store/selectors.js` file. Based on the code's logic, we can assume that `reusableBlock` will always be defined, so the chaining is added alongside `title` and `content`.

By using optional chaining, the code should gracefully handle scenarios where `title` or `content` might be undefined, potentially preventing a TypeError from being thrown.

## Testing Instructions

> **Note**
> For testing on the web, Playground's PR previewer can be used: https://playground.wordpress.net/gutenberg.html

As the crash isn't easily reproducible, there isn't a straight forward way to ensure this PR fixes it. Instead, the logic of making these change should be verified.

In both the app and on the web, we should also ensure that there are no regressions related to both synced and unsynced patterns, especially those with no titles:

* Publish a new synced or unsynced pattern with no title.
* Begin a new post and add the synced or unsynced pattern to it.
* Verify that there is no issue related to there being no title.
